lib/ansible/utils/collection_loader/_collection_finder.py

```diff
@@ -78,6 +84,101 @@ def is_python_identifier(self):  # type: (str) -> bool
 
 
 PB_EXTENSIONS = ('.yml', '.yaml')
+SYNTHETIC_PACKAGE_NAME = '<ansible_synthetic_collection_package>'
+
+
+class _AnsibleNSTraversable:
+    def __init__(self, *paths):
+        self._paths = [pathlib.Path(p) for p in paths]
+
+    def __repr__(self):
+        return '_AnsibleNSTraversable(\'%s\')' % '\', \''.join(map(to_text, self._paths))
+
+    def iterdir(self):
+        return itertools.chain(*[list(p.iterdir()) for p in self._paths if p.is_dir()])
+
+    def is_dir(self):
+        return any((p.is_dir() for p in self._paths))
+
+    def is_file(self):
+        return False
+
+    def glob(self, pattern):
+        return itertools.chain(*[list(p.glob(pattern)) for p in self._paths if p.is_dir()])
+
+    def _not_implemented(self, *args, **kwargs):
+        raise NotImplementedError('not usable on namespaces')
+
+    joinpath = __truediv__ = __truediv__ = read_bytes = read_text = _not_implemented
+
+
+class _AnsibleTraversableResources:
+    def __init__(self, package, loader):
+        self._package = package
+        self._loader = loader
+
+    def _get_name(self, package):
+        try:
+            # spec
+            return package.name
+        except AttributeError:
+            # module
+            return package.__name__
+
+    def _get_package(self, package):
+        try:
+            # spec
+            return package.__parent__
+        except AttributeError:
+            # module
+            return package.__package__
+
+    def _get_path(self, package):
+        try:
+            # spec
+            return package.origin
+        except AttributeError:
+            # module
+            return package.__file__
+
+    def _is_ansible_ns_package(self, package):
+        origin = getattr(package, 'origin', None)
+        return all(
+            (
+                origin,
+                any(
+                    (
+                        origin == SYNTHETIC_PACKAGE_NAME,
+                        os.path.basename(origin) in ('__synthetic__', '__init__.py'),
+                    )
+                )
+            )
+        )
```

Comment:
This feels excessively overengineered. Maybe simplify it a bit?

```suggestion
        if not origin:
            return False

        if origin == SYNTHETIC_PACKAGE_NAME:
            return True

        module_filename = os.path.basename(origin)
        return module_filename in {'__synthetic__', '__init__.py'}
```